### ISSUE TITLE
Deploy fails for Custom Fields on packaged objects: "Must specify a non-empty label for the CustomObject"

### ISSUE THREAD

**Issue by goto-dev-null** [forcedotcom/cli#1220]
### Summary

Custom Fields on objects that reside in packages (both managed & unmanaged/unlocked) fail to deploy, requiring an object-meta.xml file that already exists in the package.

A "workaround" is to have the full object-meta.xml file in two places, which is not very acceptable for an org-based development pattern. Or any, for that matter.

### Steps To Reproduce:

I'm sorry but I don't know if I have the ability to create an easy reproduction. I think that would require creating an unlocked package & then unpackaged source..... This is most certainly a confirmed issue though.

https://trailblazer.salesforce.com/issues_view?id=a1p4V000000KTo0QAG&title=must-specify-a-non-empty-label-for-the-customobject

1. Have a package that has an object CustomObject__c in it (i.e. the source of truth for CustomObject__c.object-meta.xml)
2. Using `sfdx force:source:deploy` to try to deploy CustomObject__c.CustomField__c

### Expected result

Deploys successfully

### Actual result

Get the error: "Get "Error:Must specify a non-empty label for the CustomObject""

### System Information

```
{
        "cliVersion": "sfdx-cli/7.118.1",
        "architecture": "win32-x64",
        "nodeVersion": "node-v14.17.4",
        "pluginVersions": [
                "@oclif/plugin-autocomplete 0.3.0 (core)",
                "@oclif/plugin-commands 1.3.0 (core)",
                "@oclif/plugin-help 3.2.3 (core)",
                "@oclif/plugin-not-found 1.2.4 (core)",
                "@oclif/plugin-plugins 1.10.1 (core)",
                "@oclif/plugin-update 1.4.0-3 (core)",
                "@oclif/plugin-warn-if-update-available 1.7.0 (core)",
                "@oclif/plugin-which 1.0.3 (core)",
                "@salesforce/sfdx-diff 0.0.6",
                "alias 1.1.10 (core)",
                "auth 1.7.1 (core)",
                "config 1.2.34 (core)",
                "generator 1.2.0 (core)",
                "mo-dx-plugin 0.3.2",
                "plugin-lightning-testing-service 1.0.1",
                "salesforcedx 52.0.0",
                "├─ data 0.4.11",
                "├─ apex 0.2.2",
                "├─ custom-metadata 1.0.12",
                "├─ user 1.3.0",
                "├─ org 1.6.6",
                "├─ schema 1.0.7",
                "├─ templates 51.5.0",
                "├─ limits 1.2.1",
                "├─ @salesforce/sfdx-plugin-lwc-test 0.1.7",
                "└─ salesforce-alm 52.0.0",
                "sfdx-cli 7.118.1 (core)",
                "sfdx-git-delta 4.9.0",
                "sfdx-git-packager 0.3.3",
                "sfdx-isdot 2.6.3 (link) C:\\Users\\jpetraglia\\repos\\isdot",
                "sfdx-jayree 3.11.2",
                "└─ @jayree/sfdx-plugin-prettier 1.0.4",
                "source 1.0.15 (core)",
                "telemetry 1.2.3 (core)",
                "texei-sfdx-plugin 1.10.1",
                "trust 1.0.7 (core)"
        ],
        "osVersion": "Windows_NT 10.0.17134"
}
```

**Comment by mshanemc** [comment]
please uninstall `salesforcedx` and try it again!  You're missing out on all kinds of new fixes.

`sfdx plugins:uninstall salesforcedx`

**Comment by goto-dev-null** [comment]
@mshanemc Thanks for the response. Unfortunately no change thought.

I'll try to give more details about the scenarios.

## Scenario 1: Unmanaged package
  1. `force-app/main/default/MyObject/MyObject__c.object-meta.xml` [and the majority of fields, validation rules, etc] exists in MyPackage.
  2. `src/main/default/MyObject/fields/MyField__c.object-meta.xml` exists in our Happy Soup [unpackaged code].
  3. Trying to deploy gives the "Must specify a non-empty label for the CustomObject".
  
Copying MyObject__c.object-meta.xml from MyPackage to the happy soup lets it deploy successfully, but obviously having our source in 2 locations is no an option.
  

## Scenario 2: Managed package
  1. Up until this issue starting randomly, SFDX used to want there to *not* be a file in the Happy Soup at `force-app/main/default/zqu__Quote__c/zqu__Quote__c.object-meta.xml` 
  2. Also up until recently, `src/main/default/zqu__Quote__c/fields/zqu__Quote__c.object-meta.xml` exists in our Happy Soup with no qualms.
  3. But now, trying to deploy gives the "Must specify a non-empty label for the CustomObject".

The only way to get the full zqu__Quote__c.object-meta.xml is to use Workbench or the Metadata API. But I really doubt we want it checked in since it's part of a managed package?


Here's my new system environment; not that this also effects our CI/CD system, which is Linux.

```
{
        "cliVersion": "sfdx-cli/7.118.1",
        "architecture": "win32-x64",
        "nodeVersion": "node-v14.17.4",
        "pluginVersions": [
                "@oclif/plugin-autocomplete 0.3.0 (core)",
                "@oclif/plugin-commands 1.3.0 (core)",
                "@oclif/plugin-help 3.2.3 (core)",
                "@oclif/plugin-not-found 1.2.4 (core)",
                "@oclif/plugin-plugins 1.10.1 (core)",
                "@oclif/plugin-update 1.4.0-3 (core)",
                "@oclif/plugin-warn-if-update-available 1.7.0 (core)",
                "@oclif/plugin-which 1.0.3 (core)",
                "@salesforce/sfdx-diff 0.0.6",
                "@salesforce/sfdx-plugin-lwc-test 0.1.7 (core)",
                "alias 1.1.10 (core)",
                "apex 0.2.8 (core)",
                "auth 1.7.1 (core)",
                "config 1.2.34 (core)",
                "custom-metadata 1.0.12 (core)",
                "data 0.6.1 (core)",
                "generator 1.2.0 (core)",
                "limits 1.2.1 (core)",
                "mo-dx-plugin 0.3.2",
                "org 1.7.0 (core)",
                "plugin-lightning-testing-service 1.0.1",
                "salesforce-alm 52.3.3 (core)",
                "schema 1.0.8 (core)",
                "sfdx-cli 7.118.1 (core)",
                "sfdx-git-delta 4.9.0",
                "sfdx-git-packager 0.3.3",
                "sfdx-isdot 2.6.4 (link) C:\\Users\\jpetraglia\\repos\\isdot",
                "sfdx-jayree 3.11.2",
                "└─ @jayree/sfdx-plugin-prettier 1.0.4",
                "source 1.0.15 (core)",
                "telemetry 1.2.3 (core)",
                "templates 52.1.0 (core)",
                "texei-sfdx-plugin 1.10.1",
                "trust 1.0.7 (core)",
                "user 1.4.0 (core)"
        ],
        "osVersion": "Windows_NT 10.0.17134"
}
```

**Comment by goto-dev-null** [comment]
@mshanemc Any more ideas?

**Comment by keirbowden** [comment]
I'm seeing this since I updated my CLI last night. I'm developing a 2GP extension to a 1GP package. I have a bunch of stuff around the 1GP package that I need to get the code to push in an unpackagedMetadata` directory. Since updating, all of the objects and fields from the unpackagedMeta directory are being duplicated in the default directory when I pull,  and I'm unable to push any updates. 

If I remove the duplicated custom object from the default directory, it tells me that I can't delete a managed component. If I add it in there I get the non-empty label error. It seems like the latest update has broken the CLIs ability to distinguish between the default and non-default directories for some metadata.

**Comment by goto-dev-null** [comment]
@keirbowden  Can you post the output of `sfdx --version` and `sfdx plugins --core`? Mentioning an upgrade makes that very relevant.

**Comment by keirbowden** [comment]
TL;DR - I can't get it to happen again.

I've been trying to reproduce this since I posted above, without success. I upgraded to the latest (7.122) and next time I pulled I noticed that a bunch of objects/fields from the 1GP had been retrieved into my default 2GP packaging directory.  Seemed to be limited to the 1GP objects that I'd added fields to in my 2GP.  Doubly weird was at least one of them was excluded via .forceignore. Thereafter I was unable to push because it either thought I was deleting a component or the same label error you got.

I then downgraded my CLI to 7.120), deleted my scratch org and created a new one. Everything pushed as expected so I ran the update command and went back to 7.122.  Regardless of what I tried after that, everything behaved as it should. I recreated everything again starting at 7.122 and I've still not seen any issues. Right now I think it may have been some combination of source tracking and the upgrade. I'm also fairly sure this scratch org was upgraded as part of the Winter 22 release which might have had an impact too. 

My current versions are:

sfdx-cli/7.122.1 darwin-x64 node-v14.18.0

@oclif/plugin-autocomplete 0.3.0 (core)
@oclif/plugin-commands 1.3.0 (core)
@oclif/plugin-help 3.2.3 (core)
@oclif/plugin-not-found 1.2.4 (core)
@oclif/plugin-plugins 1.10.1 (core)
@oclif/plugin-update 1.5.0 (core)
@oclif/plugin-warn-if-update-available 1.7.0 (core)
@oclif/plugin-which 1.0.3 (core)
@salesforce/sfdx-plugin-lwc-test 0.1.7 (core)
alias 1.1.21 (core)
apex 0.3.0 (core)
auth 1.7.2 (core)
config 1.2.40 (core)
custom-metadata 1.0.12 (core)
data 0.6.2 (core)
generator 1.2.0 (core)
limits 1.2.2 (core)
org 1.8.1 (core)
salesforce-alm 52.6.0 (core)
schema 1.0.8 (core)
sfdx-cli 7.122.1 (core)
source 1.2.2 (core)
telemetry 1.2.6 (core)
templates 52.4.0 (core)
trust 1.0.8 (core)
user 1.5.0 (core)

**Comment by goto-dev-null** [comment]
@keirbowden I'm sad that we have 1 less venue of data for troubleshooting but happy for you that it started working :P

**Comment by WillieRuemmele** [comment]
Hi @jonnypetraglia sorry for the delayed response. I'm just confirming, before digging in too much more, that if you upgrade to >7.122 like Keir did above is this issue still present?

**Comment by goto-dev-null** [comment]
Hey @WillieRuemmele Sorry for the delayed response, was actually on vacation 😎 

Just updated to 7.124.0 and the problem persists.

I do need clarification: what is the directory *supposed* to look like? I am assuming "object-meta.xml should not exist" but I've yet to have that confirmed outright here. I believe that's the conclusion Salesforce Support has come to in the case on this matter but want to confirm.

**Comment by WillieRuemmele** [comment]
Don't be sorry if you're on vacation! 🍹 

sorry to hear it's still happening... I need a little clarification on your clarifying question though...
>what is the directory supposed to look like? I am assuming "object-meta.xml should not exist"

which directory? Which `object-meta.xml` are you talking about? 

The [simple-mpd-project](https://github.com/amphro/simple-mpd-project) repository that we used when developing/testing had `CustomObjects` split across packages, and has `-meta.xml` files in each package

**Comment by goto-dev-null** [comment]
In my above examples, should `src/main/default/zqu__Quote__c/zqu__Quote__c.object-meta.xml` exist?

The sample repo is a great example, only in our case, the work in `force-app` is dependent on the `my-app` package. With this in mind, in the above repo, the meta file for the object exists in two places, `my-app/objects/MyObj__c/MyObj__c.object-meta.xml` and `force-app/main/default/objects/MyObj__c/MyObj__c.object-meta.xml`.

Why does this file exist in two places? This seems like a massive problem. It's like keeping 2 versions an Apex class. If I made jonny-app that dependent on my-app? Would there have to be a `jonny-app/objects/MyObj__c/MyObj__c.object-meta.xml` as well? And if willie-app was made to be dependent on jonny-app, that would need it to, is what I'm gathering.


Again, the issue I see is that this file needs to only exist in once place, and that only makes sense to be the package. Previously, (in the above context) we would be able to deploy `force-app` using `sfdx force:source:deploy` just fine without `force-app/main/default/objects/MyObj__c/MyObj__c.object-meta.xml` existing. Now it has to exist.

**Comment by WillieRuemmele** [comment]
@jonnypetraglia Thanks for the explanation. I now understand. I'm guessing that this file started to be required around `7.112.0` is that correct? We started using a new [library](https://github.com/forcedotcom/source-deploy-retrieve) in that release which caused quite a few issues. We knew it supported the multiple-package-development (MPD) workflow that we had just introduced in the CLI last year... but it seems like it might not be totally backwards-compatible. 

so if that is true, then I can answer your question. For older versions of the CLI, you'll only need the one version of the `-meta.xml` file like you had, and for newer ones, at least for now, you can work around it by having a `-meta.xml` in each package. 

can you confirm when you started to see this issue?

**Comment by goto-dev-null** [comment]
@WillieRuemmele The absolutely weirdest thing is that it stopped working randomly, without any upgrade. We even specifically switched to using NPM to manage SFDX so there's no background upgrading without us knowing to better catch this kind of thing.

It's not impossible that it's that version. I opened a support case on 9/14, so probably a bit before that but that looks like it would be [7.119.0](https://github.com/salesforcecli/sfdx-cli/blob/main/CHANGELOG.md#71190-2021-09-17) because 7.112.0 was released back in 7/30. I don't feel like we had this sitting for 6 weeks before opening a case, but time flies when you're having fun, so maybe.

I'm honestly stoked to hear about that new library, that seems like a pretty solid avenue to chase down. It looks like it's really cool to use too. If I remember correctly, @salesforce/plugin-source started being used recently, right? So is that plugin using the new library?

But follow up would be, what's the timeline on a fix? Obviously bugs are bugs but since the source seems to have been identified, is there any expectation that this might be resolved shortly, or should I start working on heavy automation to keep these files in sync?

**Comment by WillieRuemmele** [comment]
yeah, that library powers `plugin-source` and the new `source:beta:*` commands like `source:beta:push` and the rest that weren't initially in `plugin-source`. Eventually that library will be behind all of the `source:*` and `mdapi:*` commands 🤘 

for a timeline... that's a harder question to answer. I'll mark it as a bug and get it in our backlog.

**Comment by uip-robot-zz** [comment]
This issue has been linked to a new work item: [W-10122191](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000LjG3IYAV/view)

**Comment by shetzel** [comment]
@jonnypetraglia - I've tried for an hour to repro this and can't.  We're going to need more details and concrete repro steps for any progress on this.

I tried to repro using: https://github.com/salesforcecli/sample-project-multiple-packages

I deployed the force-app package dir first, which has the object definition for MyObj__c:
    `sfdx force:source:deploy -p force-app`
then I deployed the my-app package dir, which has a field and no object definition for the custom object in force-app:
    `sfdx force:source:deploy -p my-app`
Deploying the field was successful.

Maybe it matters how the package was deployed/installed in the target org?  The known issue does not contain enough detail about what is being retrieved and deployed.

**Comment by goto-dev-null** [comment]
@shetzel I will try my best to create a repo that is consistently reproduceable. Here's some more color in the meantime.

One different I know is that we use a manifest to deploy the unpackaged metadata, so instead of `sfdx force:source:deploy -p my-app` it would be `sfdx force:source:deploy -x manifest/primary.xml`, which introduces the whole concept of what should be in the manifest.

I think you're also onto something in that -in our situation- we don't do a `force:source:deploy` command, we would install the `force-app` package.

So the steps, roughly, would look more like:
```
sfdx force:package:install -w 901 -p 04tXXXXXXXXXXXX
sfdx force:source:deploy -x manifest/primary.xml
```

I will try to fork that repo you linked to & get it set up to be reproduceable, please keep this issue open in the meantime.

**Comment by shetzel** [comment]
I'll try using a manifest that only contains the field, but I suspect that the key to the repro is how the object was deployed/installed.  If there was a namespace associated with the object then I wonder if the field also needs to be deployed with a namespace.  Those are details I don't know so I'd have to consult with the packaging experts here.

The error in the title is suggesting that the manifest needs the `fullname` xml element to have a value and I bet it's expecting that value to match what was used in the package install.

UPDATE: I can successfully deploy using a manifest with just the field too.  If you can set the `SFDX_MDAPI_TEMP_DIR` env var before you do the package install and the deploy, check the manifests for both actions.  I bet the package install has a manifest with a fullname and the deploy manifest doesn't and that's the issue.  If that's the case then I think it might depend on the type of package being installed.

**Comment by gbutt-tr** [comment]
I ran into this issue today. It happens when you try to force:source:deploy an object folder where the CustomObject__c.object-meta.xml is missing required fields. In this case it is missing `<label>`. 

I am not sure if this is a regression or user error.

Edit: I was able to successfully deploy a partial object definition by removing the CustomObject__c.object-meta.xml from my folder and running force:source:deploy. In my case I am adding two Record Types and a Perm Set. So my source folder looks like this

```
src
  |-objects
  | |-CustomObject__c
  |   |-recordTypes
  |     |-rt1.recordType-meta.xml
  |     |-rt2.recordType-meta.xml
  |-permissionsets
    |-ps1.permissionset-meta.xml
```

**Comment by goto-dev-null** [comment]
I'm still encountering this issue, I just haven't had the time to set up a good repo, upon which I'll hope this issue can be reopened.

But to @gbutt-tr, I wish you had read this thread more carefully. It's not user error, I'm using the exact same meta files moved through different folders that work in one place and do not in another. I am not manually editing the files to remove tags. And I don't know how rephrasing "Must specify a non-empty label for the CustomObject" as "It happens when you try to force:source:deploy an object folder where the CustomObject__c.object-meta.xml is missing required fields" is supposed to add any clarity.

No, it's not missing a label, the file is not supposed to exist there, that's the point. It just happens to pick 'label' as the first tag to complain about, but the contents of the file -up until this point- was
```
<?xml version="1.0" encoding="UTF-8"?>
<CustomObject xmlns="http://soap.sforce.com/2006/04/metadata">
    <enableEnhancedLookup>false</enableEnhancedLookup>
</CustomObject>
```

Can you please be more specific? I've found that SFDX retrieves and deploys **very** differently for manifest vs component type vs path. Especially for objects. What **exactly** is the deploy command you're using, flags and all? As I already stated, we are using a manifest.


The problem has actually gotten worse. We moved a custom object that has a MasterDetail on it from our soup into the package, so now when I try to copy CustomObject__c.object-meta.xml into our soup to get around the problem, it complains about the sharing model because the soup doesn't have the Primary Key, even though it exists in the org. Deployment for all things on it that sit outside of the package is literally broken for this object.

EDIT: I am now actively investigating this. I created a skeleton project to reproduce but was unable to, yet I switch to our production repo and it still occurs. I am going to whittle away our manifest until I identify the problem, but that could take some time so any other ideas are appreciated.

A thought I had, could this have to do with having CustomObject__c.workflow-meta.xml in the repo but not CustomObject__c.object-meta.xml? It's the only other place our CustomObject__c is named in the manifest that causes the failure.

**Comment by goto-dev-null** [comment]
Okay, so I finally found the source of the problem, and it's not what's in the title of the ticket.

It's because we have the wildcard in our manifest:
`
	<types>
		<members>*</members>
		<name>CustomObject</name>
	</types>
`
Does this wildcard go by folder names instead of object-meta.xml files? That would majorly suck.

Project will shell script to reproduce but here's also the directory structure:
![image](https://user-images.githubusercontent.com/2322497/147791322-e07e5384-75a9-4b6a-a486-b0a90d64dac1.png)

https://github.com/jonnypetraglia/sfdx-error1

**Comment by vc4u** [comment]
For me, this the workaround that worked:

- You need to delete the `<NAMESPACED-OBJECT>.object-meta.xml` files from force-app's objects directory.
- Also remove the references for same objects from `.sfdx/orgs/<ORG-NAME>/sourcePathInfos.json` for files ending with `.object-meta.xml`.

Source: https://salesforce.stackexchange.com/questions/194509/how-to-push-fields-on-packaged-objects-with-sfdx

**Comment by goto-dev-null** [comment]
@vc4u Your comment is not relevant for the details in my latest comment. The issue is not the meta file itself but the entry in the manifest, and "sourcePathInfos.json" is for source tracking while this ticket is explicitly talking about `force:source:deploy` which -to my knowledge- does not use source tracking.

**Comment by shetzel** [comment]
@jonnypetraglia - thanks for the repo!  I see what you're doing now.  You're right that the problem section of your manifest is the wildcarded custom objects.  If you remove it the deploy is successful.  If you only have the CustomObject wildcard in the manifest it fails in the same way (because it needs the fullname in the manifest).  The wildcard does not care about folder names.  The error is because you're deploying the packaged CustomObjects (i.e., the objects in `package/main/default/objects/*`) along with unpackaged metadata.  Since those objects are part of the installed package it expects the manifest to have a fullname element with a value of "GithubTest".  Like you said, if you remove the wildcarded custom objects section from the manifest it works fine because it's deploying all unpackaged metadata.  If you keep it in, then you're mixing unpackaged metadata with packaged metadata and I'm not sure if that's allowed.  I'll have to run this by a few people and get back to you.
Thanks for all your effort on this!  It's made a huge difference in getting this solved.

**Comment by goto-dev-null** [comment]
@shetzel Thanks for the super detailed explanation!! I appreciate it!

**Comment by shetzel** [comment]
@jonnypetraglia - Sorry for such a delayed followup.  I think the bug here is that your deploy command is acting on a directory within your project that is not defined as a package directory within your `sfdx-project.json`.  Unless the `source:deploy` command is run with the sourcepath command flag it should scope the manifest (either generated or provided) to the package directories defined in `sfdx-project.json`.  We'll fix that bug.  Thanks again for your sample repo and patience!  When the fix is in the CLI (the source plugin) we'll post here and close this issue.

**Comment by shetzel** [comment]
I started digging into this today and discovered that I was inaccurate with my post from late January.  The `source:deploy` command is properly scoping to the defined package directories when the `--metadata` and `--manifest` flags are used.  This is really about how the Metadata API interprets the metadata being deployed in combination with the manifest when it contains a CustomObject wildcard and the metadata extends a packaged CustomObject.  When the Metadata API sees `CustomObject: *` and a CustomObject file (even though it only contains just the field you want to deploy) and that CustomObject is part of a package it will expect the namespace to be defined in the manifest.  So the moral of the story here is that when you're deploying extensions to a packaged CustomObject you have to be very careful and specific with your manifest contents.

I'm closing this issue as the CLI is behaving correctly.

**Comment by teodora-raileanu** [comment]
@shetzel I am currently facing the same issue, can you elaborate a bit on the solution that you have found?

**Comment by shetzel** [comment]
Without knowing all the details of your specific case I'm not sure what more to say.  Essentially, you can't deploy packaged metadata with unpackaged metadata.  Make them separate deploy commands.  It might be best to create a new issue with all your details and maybe reference this one.  Definitely include output of `sfdx version --verbose`, what the deployment manifest looks like, the exact command and parameters (not necessarily the param values), and some details on the CustomObject and CustomFields.  If it's exactly like this issue then you want to be very specific in your manifest with the CustomObject and CustomFields and avoid using a wildcard.

**Comment by togu09** [comment]
We are currently experiencing this issue when trying to push a new custom object with some fields to a scratch org. I've tried different versions of the CLI and it's working fine in 7.122.0, not working in 7.150.0 and 7.162.0.

"Error   ACTM_AccountTeamMember__c                                Must specify a non-empty label for the CustomObject"

I'm however getting the error when running the sfdx force:source:push command. The metadata is in an unlocked package and we're pushing multiple unlocked packages and unpackaged metadata that resides in different folders in the same push. It was working fine for us on version 7.122 and 7.150 until we introduced this new object.
